pkg/tls/config.go

```diff
@@ -45,54 +42,72 @@ var (
 		"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305":    tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
 		"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305":  tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
 	}
+
+	validCiphers13 = map[string]uint16{
+		"TLS_AES_128_GCM_SHA256":       tls.TLS_AES_128_GCM_SHA256,
+		"TLS_AES_256_GCM_SHA384":       tls.TLS_AES_256_GCM_SHA384,
+		"TLS_CHACHA20_POLY1305_SHA256": tls.TLS_CHACHA20_POLY1305_SHA256,
+	}
 )
 
-func BaseTLSConfig() (*tls.Config, error) {
-	// Get configured minimal tls version
-	TLSMinVersion, err := lookupTLSVersion()
+func baseTLSConfig(minVersion, ciphers string) (*tls.Config, error) {
+	version, err := validatedMinVersion(minVersion)
 	if err != nil {
-		return nil, fmt.Errorf("error while configuring minimal TLS version: %s", err)
+		return nil, err
 	}
-	// Get configured tls ciphers
-	TLSCiphers, err := lookupTLSCiphers()
+	cipherSuites, err := validatedCiphers(ciphers, version)
 	if err != nil {
-		return nil, fmt.Errorf("error while configuring TLS ciphers: %s", err)
+		return nil, err
 	}
-
 	return &tls.Config{
-		PreferServerCipherSuites: true,
-		MinVersion:               TLSMinVersion,
-		CipherSuites:             TLSCiphers,
+		MinVersion:   version,
+		CipherSuites: cipherSuites,
 	}, nil
 }
 
-func lookupTLSVersion() (uint16, error) {
-	tlsVersionKeys := getKeysFromMap(tlsVersions)
-	settingsTLSMinVersion := settings.TLSMinVersion.Get()
-	if val, ok := tlsVersions[settingsTLSMinVersion]; ok {
+func validatedMinVersion(version string) (uint16, error) {
+	if val, ok := validVersions[strings.TrimSpace(version)]; ok {
 		return val, nil
 	}
-	return 0, fmt.Errorf("invalid minimal TLS version [%s], must be one of: %s", settingsTLSMinVersion, strings.Join(tlsVersionKeys, " "))
+	valid := keysFromMap(validVersions)
+	sort.Strings(valid)
+	return 0, fmt.Errorf("unsupported minimal TLS version %s, must be one of: %s", version, strings.Join(valid, ", "))
 }
 
-func lookupTLSCiphers() ([]uint16, error) {
-	tlsCipherSuitesKeys := getKeysFromMap(tlsCipherSuites)
-	settingsTLSCiphers := settings.TLSCiphers.Get()
-	sliceTLSCiphers := strings.Split(settingsTLSCiphers, ",")
+func validatedCiphers(ciphers string, version uint16) ([]uint16, error) {
+	split := strings.Split(ciphers, ",")
+	if version == tls.VersionTLS13 {
+		return validatedCiphersHelper(split, validCiphers13)
+	}
+	return validatedCiphersHelper(split, union(validCiphers10to12, validCiphers13))
+}
 
-	var TLSCiphers []uint16
-	for _, TLSCipher := range sliceTLSCiphers {
-		val, ok := tlsCipherSuites[strings.TrimSpace(TLSCipher)]
+func validatedCiphersHelper(ciphers []string, validCiphers map[string]uint16) ([]uint16, error) {
```

Comment:
nit:
I don't love calling things helper functions.
```suggestion
func getCipherValues(ciphers []string, validCiphers map[string]uint16) ([]uint16, error) {
```

Reply:
Yes, adjusted!